pandas/tseries/offsets.py

```diff
@@ -261,16 +261,12 @@ def apply_index(self, i):
         # relativedelta/_offset path only valid for base DateOffset
         if (self._use_relativedelta and
             set(self.kwds).issubset(relativedelta_fast)):
+
             months = ((self.kwds.get('years', 0) * 12
                         + self.kwds.get('months', 0)) * self.n)
             if months:
-                base = (i.to_period('M') + months).to_timestamp()
-                time = i.to_perioddelta('D')
-                days = i.to_perioddelta('M') - time
-                # minimum prevents month-end from wrapping
-                day_offset = np.minimum(days,
-                                        to_timedelta(base.days_in_month - 1, unit='D'))
-                i = base + day_offset + time
+                shifted = tslib.shift_months(i.asi8, months)
+                i = i._constructor(shifted)
```

Comment:
This will lose the `name` of the `Index` - I think you want to use `_shallow_copy`:

```python

In [57]: df.index._constructor(df.index)
Out[57]: 
Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
            19],
           dtype='int64')

In [58]: df.index._shallow_copy(df.index)
Out[58]: 
Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
            19],
           dtype='int64', name='hi')
```
(there's a broader issue about the constructor keeping the name)